### ISSUE TITLE
A proposal for fixing CSV downloads in notify

### ISSUE THREAD

**Issue by currycoder** [alphagov/notifications-python-client#164]
**The Problem**

GOVUK notify (https://www.notifications.service.gov.uk/) is a web service we are using to send emails to external partners. For our use case, we need to send these partners spreadsheet files (CSVs) in an agreed format for them to process and send back to us. GOVUK notify offers a mechanism for file uploads/downloads which we would like to use for this purpose.

Unfortunately, CSV files uploaded as part of a notify email message end up being saved as TXT files. This means that our partners download these files as TXT files, which is against our agreed process and is causing us problems.

The problematic bit of code lies in GOVUK's document-download-api web service, here: https://github.com/alphagov/document-download-api/blob/master/app/upload/views.py#L18

The `get_mime_type()` utility acts based on the contents of a file only (at this point in the flow, the service does not have access to file name/extension) - and so it identifies the file as a TXT file instead of as a CSV.

**Proposed Solution**

We need to provide a solution here which works both for our particular situation and for existing notify users (who may have worked around this problem in their own ways).

The proposal is for us to adjust `notifications-python-client`, `notifications-api` and `document-download-api` so that we can optionally pass the correct file mimetype through to the logic that saves the file.

This would require the following changes:

- The `prepare_upload()` utility in `notifications-python-client` would need to optionally take a mimetype kwarg: https://github.com/alphagov/notifications-python-client/blob/ddf6b1afb4b8bb22badc835a93c4f95a33eb2ccf/notifications_python_client/utils.py#L6. This should result in the `mimetype` being added to the return dictionary e.g. `{'data': '<base64>', 'mimetype': 'text/csv'}`.
- The API endpoint for sending email notifications on `notifications-api` would need adjusting to accept mimetype being passed through in JSON objects representing files: https://github.com/alphagov/notifications-api/blob/master/app/v2/notifications/post_notifications.py#L117
- The method that `notifications-api` uses to upload documents to `document-download-api` would need adjusting to accept and forward mimetype as part of the POST body. https://github.com/alphagov/notifications-api/blob/d88b20beeca799f920eb84fb1695f336e4fb5a29/app/clients/document_download.py#L29
- The API endpoint for uploading documents on `document-download-api` would need adjusting to optionally accept mimetypes in the POST body: https://github.com/alphagov/document-download-api/blob/master/app/upload/views.py#L14 This function would need to validate the mimetype by ensuring that it is in the whitelist and also ensuring that this mimetype is valid given the mimetype detected from the file content. This would then be used in the part that saves the file: https://github.com/alphagov/document-download-api/blob/master/app/upload/views.py#L32

The rollout for these changes would need to be back to front - e.g. changes for `document-download-api` and then `notifications-api` and finally `notifications-python-client`. This would hopefully take the form of one pull request for each repository. These changes are purely additive/optional and not BC breaking so hopefully deployment should be simple.

**Alternative Solutions**

- It could be possible to simply improve the mimetype auto detection in the `document-download-api` upload endpoint. However, this would be fiddly to do reliably and this could present a backwards compatibility breaking change for clients that assume that CSVs will be mis-identified as TXT files.
- `notifications-python-client` could be adjusted to instead send the file extension (`csv`) or whole file name (`myfile.csv`) to make things simpler for clients. This would add some complexity somewhere later in the chain to resolve this to a mimetype.

**Further Questions**

- Do all of the other notification client libraries need adjusting as part of this work? It would be preferable to get this working for the python client first, from my perspective.
- The documentation probably needs some adjustment as part of this work. Should this also be done in PRs?  Are the docs also under version control?  Where?

**Comment by currycoder** [comment]
> This function would need to validate the mimetype by ensuring that it is in the whitelist and also ensuring that this mimetype is valid given the mimetype detected from the file content.

I've assumed here that we cannot just respect the mimetype that the client sends as our only source of truth - e.g. the situation where somebody sends a .zip file but specifies that it is a`txt`.  My thought is that we will need some sort of mapping to say "an explicitly specified CSV mimetype is acceptable if we detect this file to be TXT".

Any comments on this particular train of thought would be apprecitated.

**Comment by idavidmcdonald** [comment]
Hi,

We've discussed this via slack briefly but is here the public reply. Once again, thanks for your help on this, it's great to see contributions and it's much much appreciated!

I've discussed this with another developer on the team and think we want the following, which is a small change from your design but still pretty similar.

## Proposed solution
An additional argument is added to `prepare_upload`. The argument should be called `is_csv` and is an optional boolean to indicate whether the file that is being uploaded is a CSV file. We think the following is a first stab at an explanation for this argument to go in the API docs:

```
If you upload a .csv file, set this to True to ensure that your user downloads it as a .csv file rather than a .txt file. If the file you upload is not a CSV file then we will ignore this.
```

We prefer using this argument to either something like `file_extension`, `file_name` or `content_type` as it's much simpler for the user to understand if they need to use it (as it won't be relevant for most) and should hopefully keep confusion to a minimum (no need for our API users to go reading round about mimetypes if they want to upload a PDF file for example).

The rest of your solution will stay roughly the same in terms of arguments being passed starting from the API client, to the API, to document download API.

### Content disposition
I've done some manual testing of setting the mimetype to `text/csv` for files I've uploaded to S3. I've then tested what happens when I open them in document download in Chrome and Safari.

In Chrome, the response is to open Google Sheets.
In Safari however the response is still to view the csv file in browser as plaintext, rather than downloading.

I believe this is not the desired behaviour for Safari (and maybe other browsers that I haven't tested), at least it sounds like not in your case as you would want the user to download the file? Is that correct?

A proposed solution to this is that we add the `Content-Disposition: attachment` header to responses for files of `text` mimetype (both plain and csv) so that all text files are downloaded, but we allow `application` mimetypes to still let the browser decide. We generally are keen to let the browser decide and not force a download (we avoided this in our original design as think it adds confusion for users, in particular mobile users on android when dealing with downloads) where possible but think that in the case of a .txt or .csv file, it makes more sense to download it. Thoughts? This setting of content disposition could also come as a separate piece of work so not all of these changes need to happen at once.

## Test cases
We put together a few test cases, of the following format:

file format that python-magic believes the file is, the value of the `is_csv` argument and the resulting mimetype that gets saved in S3.

(txt, True) -> text/csv
(txt, False) -> text/plain
(txt, None) -> text/plain
(pdf, False) -> application/pdf
(pdf, None) -> application/pdf
(pdf, True) -> application/pdf (Note, no 4xx raised the for user, we just ignore the CSV argument)

## Long run solution
We also spent some time discussing that in the long run we think we'd want to add an argument (maybe optional, maybe not) for a filename to the `prepare_upload` method. Then content type could be taken from that and we also could then give the user a better filename to download the file as (at the moment it's a UUID). However, we think this is too big and complex to do to solve this bug in particular.

## Responses to your other questions
1) Let's start with just the python client. In time we will want to do all the other API clients but it will be great to have you/other python users test out the feature first and then we can extend it in time to the other clients. You are more than welcome to do the work on the other API clients but it's not expected, contributing this feature to the python client should take up a fair bit of your time to start with!

2) Yes please on updating the docs (for the python client only). The documentation can be found here https://github.com/alphagov/notifications-python-client/blob/master/DOCUMENTATION.md. We will also want a relevant change to the CHANGELOG and version bump to the API client. You can find what is needed looking in our pull request template https://github.com/alphagov/notifications-python-client/blob/master/.github/PULL_REQUEST_TEMPLATE.md

**Comment by currycoder** [comment]
Thanks for this response David.  As discussed in slack, I appreciate that the solution you are proposing here is reduced in scope and provides a reasonable fix for the problem we are experiencing without complicating the overall experience too much for notify users.

I've gone ahead and popped up an initial PR for the `document-download-api` part of the work, here: https://github.com/alphagov/document-download-api/pull/122/files

As for your section on content disposition, this isn't something that I had anticipated.  Your plan here sounds reasonable - what we can do is put up a separate PR for this work in order to keep PRs small and easy to review.  I'll look at working on this in due course as part of this string of PRs.

**Comment by idavidmcdonald** [comment]
Closing this out as we believe this issue is now resolved in 5.6.0 of the python client. Big thanks to @currycoder and @alixedi for taking the lead on this!

![image](https://user-images.githubusercontent.com/7228605/82672770-b52dca00-9c38-11ea-8160-76b360af2034.png)
